### PR TITLE
WIP: Migrate to lambda

### PR DESCRIPTION
**Pull request by jwalton3141**
This PR begins the migration from DiAGRAM as a shiny application, to DiAGRAM as a JS frontend and AWS Lambda + R backend

### CODE REVIEW

**Comment by DavidUnderdown** [comment]
Thanks Jack

I just want to double check with a couple of people that we're happy doing this in one fell swoop and that there's nothing obviously in need of any tweaks.

On a very cursory look through I've noticed that the licence file has ended up with:

`<<<<<<< HEAD
Proprietary

Do not distribute outside of Jumping Rivers Ltd.
=======`

added to the top of it which should probably be edited back out before this goes into our repo.

**Comment by jwalton3141** [comment]
Hi @DavidUnderdown.

Good catch on the LICENSE: I'd erroneously migrated that over from our internal repository!

I realise that pushing through a large number of changes in this manner is not ideal. If you would consider it more appropriate, we could of course simply start from a clean repository, and leave the current repository as it exists now.

The codebase has changed so signficantly in the migration from shiny -> {plumber} + JS -> AWS Lambda + JS, that there is almost no shared history left with the original project.

**Comment by DavidUnderdown** [comment]
Hi @jwalton3141 

Yeah, it's an interesting one, there are advantages to retaining the plain DiAGRAM name, but as you say it's such a major change in some ways a new repo would make sense.

One other thing I'm now wandering about is whether there's any way of updating the shinyapps versions to try and add something to direct people to the new version, though I do want to keep https://nationalarchives.shinyapps.io/tna_gui/, https://nationalarchives.shinyapps.io/DiAGRAM-dev/ and https://nationalarchives.shinyapps.io/DiAGRAM/ accessible to maintain the history (plus various things like recorded webinars and journal articles have references to them so it would be quite bad if they just disappeared), though I'm envisaging dropping to the free tier since people shouldn't actually be creating models in those versions once we have the new one live. I don't know if you or @jamieRowen have thoughts on that?

File: api/src/R/checks.R [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#' @param x Typically single response node, may be a numeric or character
+#' vector, or a list of such objects.
+#' @param options named vector (or list) of allowed options for a response
+check = function(x, ...) UseMethod("check")
+
+check.default = function(x, ...) return(invisible(NULL)) # nolint
+
+check.technical_skills = function(x, options, ...) { # nolint
+  assertthat::assert_that(is.character(x))
+  assertthat::assert_that(length(x) == 10)
```
Comment (sparkhi) [inline api/src/R/checks.R:12]: A lot of assertions in this file seem to be comparing with numbers, which, to me indicates that there must be a business logic reason for the fixed values. Should these be defined as constants somewhere?

**Comment by jwalton3141** [comment]
If the main motivation of merging the changes into _this_ repo would be to retain the concise "DiAGRAM" name, I wonder what consequences there would be of simply renaming this repository to eg. DiAGRAM-poc (or DiAGRAM-old, DiAGRAM-alpha, ...), and then adding these changes to a new repository, using the freed DiAGRAM name? It's possible that @sjimellor-jr might have some insights and or opinions on this.

Regarding linking from the shinyapps deployments, to the "new" deployment, once live: I'm sure that this will be possible. However, I'm unfamiliar with the shinyapps deployment and so wouldn't know how to achieve this myself.

File: api/src/R/checks.R (hunk 1 above)
Comment (DavidUnderdown) [reply api/src/R/checks.R:12]: Thanks @sparkhi: in the particular example I see there I know that we currently have 10 skills listed for input in the technical skills question, so if we changed that at some point (not impossible as currently its based on the DigCurv framework and DPC  have recently introduced a new skills framework) then the number might change. I guess that if the number is only used in one place there may not be much advantage in setting a constant versus just declaring them in the test.

**Comment by jwalton3141** [comment]
One more thing: the majority of the changes made in the `api/` directory were actually made by @jamieRowen when he developed the original {diagramAPI} plumber package.

My contributions to `api/` are largely the Dockerfile, and the custom serialisation / deserialisation required for AWS API Gateway, and the few {lambdr} bits. Similarily, from what I understand, the majority of the development within `app/` was performed by @tim347.

_However_, as the {diagramAPI} package (@jamieRowen), and the frontend code (@tim347) were separate repositories hosted as a private repository within Jumping Rivers' GitLab account, their significant contributions to this work are lost in this migration. @jamieRowen, @tim347, we should meet next week to discuss how we can perform this migration in a way that upholds and reflects your contributions. If we're sticking with this repo, it might be simplest if we simply undo the changes and have you commit the bits you contributed. I'll think on this over the weekend.

**Comment by DavidUnderdown** [comment]
Yes, I think we've probably skipped a pull request (or two) to get the previous round of changes in.

**Comment by jwalton3141** [comment]
I'm going to close this pull request and start a new one in which @jamieRowen and @timbrock's commits are added by themselves.

I _wasn't_ planning on putting their commits through separate pull-requests: the works are inexplicably linked. However, @DavidUnderdown, if you would prefer separate PR's for integrating the frontend and backend changes in, I can do that. Just let me know your preference. Thanks

**Comment by DavidUnderdown** [comment]
I think you're right, those changes together make sense as a unit

**Pull request closed without merging**